File: src/main/java/org/kurator/validation/actors/SciNameAssembler.java
package org.kurator.validation.actors;

import java.util.Map;

import org.gbif.dwc.terms.DwcTerm;
import org.kurator.akka.data.SpacedStringBuilder;
import org.kurator.akka.data.Util;
import org.kurator.validation.exceptions.MissingFieldException;

public class SciNameAssembler {

    public final String GENUS_FIELD_NAME;
    public final String SUBGENUS_FIELD_NAME;
    public final String SPECIFIC_EPITHET_FIELD_NAME;
    public final String VERBATIM_TAXON_RANK_FIELD_NAME;
    public final String TAXON_RANK_FIELD_NAME;
    public final String INFRASPECIFIC_EPITHET_FIELD_NAME;

    public SciNameAssembler(
            String genusFieldName, 
            String subgenusFieldName,
            String specificEpithetFieldName, 
            String verbatimTaxonRankFieldName,
            String taxonRankFieldName, 
            String infraspecificEpithetFieldName
    ) {
        GENUS_FIELD_NAME                 = genusFieldName;
        SUBGENUS_FIELD_NAME              = subgenusFieldName;
        SPECIFIC_EPITHET_FIELD_NAME      = specificEpithetFieldName;
        VERBATIM_TAXON_RANK_FIELD_NAME   = verbatimTaxonRankFieldName;
        TAXON_RANK_FIELD_NAME            = taxonRankFieldName;
        INFRASPECIFIC_EPITHET_FIELD_NAME = infraspecificEpithetFieldName;
    }

    public static SciNameAssembler newDwCSciNameAssembler() {
        return new SciNameAssembler(
                "genericEpithet",
                DwcTerm.subgenus.simpleName(),
                DwcTerm.specificEpithet.simpleName(),
                DwcTerm.verbatimTaxonRank.simpleName(),
                DwcTerm.taxonRank.simpleName(),
                DwcTerm.infraspecificEpithet.simpleName());
    }

    public String assembleName(Map<String, String> record) throws Exception {

        return assembleName(
                record.get(GENUS_FIELD_NAME),
                record.get(SUBGENUS_FIELD_NAME),
                record.get(SPECIFIC_EPITHET_FIELD_NAME),
                record.get(VERBATIM_TAXON_RANK_FIELD_NAME),
                record.get(TAXON_RANK_FIELD_NAME),
                record.get(INFRASPECIFIC_EPITHET_FIELD_NAME)
        );
    }

    public String assembleName(
            String genus, 
            String subgenus,
            String specificEpithet, 
            String verbatimTaxonRank, 
            String taxonRank,
            String infraspecificEpithet
    ) throws Exception {

        if (!Util.hasContent(genus)) {
            throw new MissingFieldException(
                "SciNameAssembler",
                GENUS_FIELD_NAME
            );
        }

        if (Util.hasContent(infraspecificEpithet)
                && !Util.hasContent(specificEpithet)) {
            throw new MissingFieldException("SciNameAssembler",
                    SPECIFIC_EPITHET_FIELD_NAME, 
                    "if " + INFRASPECIFIC_EPITHET_FIELD_NAME + " provided"
            );
        }

        return new SpacedStringBuilder()
                .append(genus)
                .append(subgenus)
                .append(specificEpithet)
                .append(Util.hasContent(verbatimTaxonRank) ? verbatimTaxonRank : taxonRank)
                .append(infraspecificEpithet)
                .toString();
    }
}

File: packages/kurator_names/SciNameAssembler.py
from StringUtils import has_content
from StringUtils import SpacedStringBuilder
from DarwinCore import DarwinCore

class SciNameAssembler(object):

    def __init__(
        self,
        genus_field_name                 = DarwinCore.field_names['GENUS'],
        subgenus_field_name              = DarwinCore.field_names['SUBGENUS'],
        specific_epithet_field_name      = DarwinCore.field_names['SPECIFIC_EPITHET'],
        verbatim_taxon_rank_field_name   = DarwinCore.field_names['VERBATIM_TAXON_RANK'],
        taxon_rank_field_name            = DarwinCore.field_names['TAXON_RANK'],
        infraspecific_epithet_field_name = DarwinCore.field_names['INFRASPECIFIC_EPITHET']
    ):
        self._genus_field_name                 = genus_field_name
        self._subgenus_field_name              = subgenus_field_name
        self._specific_epithet_field_name      = specific_epithet_field_name
        self._verbatim_taxon_rank_field_name   = verbatim_taxon_rank_field_name
        self._taxon_rank_field_name            = taxon_rank_field_name
        self._infraspecific_epithet_field_name = infraspecific_epithet_field_name

    def assemble_name(
            self,
            genus,
            subgenus,
            specific_epithet,
            verbatim_taxon_rank,
            taxon_rank,
            infraspecific_epithet
    ):

        if not has_content(genus):
            raise Exception('SciNameAssembler requires value for ' +
                            self._genus_field_name)

        if not has_content(infraspecific_epithet) and not has_content(specific_epithet):
            raise Exception('SciNameAssembler requires values for ' +
                            _specific_epithet_field_name +
                            ' if ' + _infraspecific_epithet_field_name + ' provided')

        return str(SpacedStringBuilder()
                   .append(genus)
                   .append(subgenus)
                   .append(specific_epithet)
                   .append(verbatim_taxon_rank if has_content(verbatim_taxon_rank) else taxon_rank)
                   .append(infraspecific_epithet)
                   )

    def assemble_name(self, record):
        """
        :type record: dict
        """
        return assemble_name(
            record[self._genus_field_name],
            record[self._subgenus_field_name],
            record[self._specific_epithet_field_name],
            record[self._verbatim_taxon_rank_field_name],
            record[self._taxon_rank_field_name],
            record[self._infraspecific_epithet_field_name]
        )
